Data-Structures/HAMT.md

```diff
@@ -1,5 +1,103 @@
-# [WIP] Hash-Array Mapped Trie
+# IPLD HAMT Spec 
 
-This specifies a standardized hash-array mapped trie on IPLD Data Model v1.
+Good reading:
+- https://blog.acolyer.org/2015/11/27/hamt/
+- https://michael.steindorfer.name/publications/oopsla15.pdf
 
-TODO: write this spec.
\ No newline at end of file
+
+The HAMT is a Key Value store implemented with a 256ary hash keyed trie. The
+basic structure in it is a `Node`. The node is as follows:
+
+```go
+type Node struct {
+	Bitfield Bitfield
+	Pointers []*Pointer
+}
+```
+
+The `Node` is serialized as a cbor object (major type 5), with the bitfield
+serialized as a cbor major type 2 byte array. The Bitfield field uses `bf` as
+its object key, and the Pointers array uses `p` as its object key.
+
+```go
+type Pointer struct {
+	KVs []*KV
+	Link Cid
+}
+```
+
+The `Pointer` is also serialized as a cbor object (major type 5), with the KVs
+field serialized as a major type 4 array of the `KV` objects, and the Link
+field serialized as an [ipld dag-cbor Cid](https://github.com/ipld/specs/blob/master/Codecs/DAG-CBOR.md#link-format).
+
+```go
+type KV struct {
+	Key string
+	Value Anything
+}
+```
+
+The `KV` is serialized as a cbor array (major type 4) with the 'key' field
+serialized as a cbor string (major type 3) (TODO: should this just be major
+type 2? its probably good to support arbitrary bytes as keys) and placed in the
+zero'th position of the array, and the value serialized { in some way } and
+placed in array position 1.
+
+## Lookup
+
+To look up a value in the HAMT, first hash the key using a 128 bit murmur3 hash.
+Then, for each layer take the first W bits of the hash, and use that to compute
+the index for your key, as follows:
```

Comment:
Isn't `W` always a single byte?

Reply:
if it's fixed to an arity of 256, but I'm wondering why 256 is chosen here, is it mainly for the ease of accessing the hash in 8-bit chunks?

Reply:
I think we cap number of links in protobuf nodes at ~170 so I think the conservative approach was taken and the number of links was kept roughly similar.

Reply:
Yeah, wanted to make it a number that would result in acceptable maximum node sizes, but 256 specifically was chosen simply because it makes reading the next index off the hash easy.

Reply:
So, there is a pathological case here: large keys. Should we impose a size limit (256 bytes?) as most filesystems do?

Reply:
I think a keysize limit is worthwhile, though that should probably be left up to the application